pkg/agent/plugin/nodeattestor/devid/devid.go

```diff
@@ -0,0 +1,376 @@
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "missing configurable: %v", err)
+	}
+
+	// Create initial internal configuration
+	intConf := &config{
+		trustDomain: req.GlobalConfig.TrustDomain,
+		devicePath:  extConf.DevicePath,
+
+		// If Attestation Key is configured, it is assumed that the user wants to verify DevID residency
+		checkDevIDResidency: akConfigured(extConf),
+	}
+
+	// Load DevID files
+	err = loadDevIDFiles(extConf, intConf)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to load DevID files: %v", err)
+	}
+
+	// Load Attestation Key files (if configured)
+	if intConf.checkDevIDResidency {
+		err = loadAKFiles(extConf, intConf)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "unable to load attestation key files: %v", err)
+		}
+	}
+
+	p.setConfig(intConf)
+
+	return &plugin.ConfigureResponse{}, nil
+}
+
+func (p *Plugin) GetPluginInfo(ctx context.Context, req *plugin.GetPluginInfoRequest) (*plugin.GetPluginInfoResponse, error) {
+	return &plugin.GetPluginInfoResponse{}, nil
+}
+
+func (p *Plugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+func (p *Plugin) getConfig() *config {
+	p.m.Lock()
+	defer p.m.Unlock()
+	return p.c
+}
+
+func (p *Plugin) setConfig(c *config) {
+	p.m.Lock()
+	defer p.m.Unlock()
+	p.c = c
+}
+
+func (p *Plugin) setPluginConfigDefaults(config *Config) {
+	if config.DevicePath == "" {
+		config.DevicePath = defaultDevicePath
+		p.log.Info("tpm_device_path is not set, using default: %q", defaultDevicePath)
+	}
+}
+
+func decodePluginConfig(hclConf string) (*Config, error) {
+	extConfig := new(Config)
+	if err := hcl.Decode(extConfig, hclConf); err != nil {
+		return nil, err
+	}
+
+	return extConfig, nil
+}
+
+func validatePluginConfig(c *Config) error {
+	// DevID certificate, public and private key are always required
+	switch {
+	case c.DevIDCertPath == "":
+		return fmt.Errorf("devid_cert_path is required")
+
+	case c.DevIDPrivPath == "":
+		return fmt.Errorf("devid_priv_path is required")
+
+	case c.DevIDPubPath == "":
+		return fmt.Errorf("devid_pub_path is required")
+	}
+
+	// Attestation private and public keys are not required but if one is
+	// provided the other one is also needed.
+	switch {
+	case c.AKPrivPath == "" && c.AKPubPath == "":
+		return nil
+
+	case c.AKPrivPath == "":
+		return fmt.Errorf("ak_priv_path is required if ak_pub_path is provided")
+
+	case c.AKPubPath == "":
+		return fmt.Errorf("ak_pub_path is required if ak_priv_path is provided")
+	}
+
+	return nil
+}
+
+func akConfigured(c *Config) bool {
+	return c.AKPubPath != "" && c.AKPrivPath != ""
+}
+
+func loadDevIDFiles(c *Config, info *config) error {
+	devIDCertBytes, err := ioutil.ReadFile(c.DevIDCertPath)
+	if err != nil {
+		return fmt.Errorf("cannot load certificate: %w", err)
+	}
+
+	info.devIDCert, err = x509.ParseCertificate(devIDCertBytes)
+	if err != nil {
+		return fmt.Errorf("cannot parse certificate: %w", err)
```

Comment:
Oh, I intentionally avoided writing "DevID certificate" in that error message because the calling function already sets the context. So for example, the full error message would be:

```
unable to load DevID files: cannot parse certificate: [parsing error description]
```

Do you think that it is clear enough? or would you like to also mention that it is the DevID certificate in the inner error?